taskvine/src/manager/vine_mount.c

```diff
@@ -0,0 +1,37 @@
+
+#include "vine_mount.h"
+
+#include <string.h>
+#include <stdlib.h>
+
+struct vine_mount * vine_mount_create( struct vine_file *file, const char *remote_name, vine_file_flags_t flags, struct vine_file *substitute )
+{
+	struct vine_mount *m = malloc(sizeof(*m));
+	m->file = file;
+	if(remote_name) {
+		m->remote_name = strdup(remote_name);
+	} else {
+		m->remote_name = 0;
+	}
+	m->flags = flags;
+	m->substitute = substitute;
+
+	/* Add a reference each time a file is connected. */
+	m->file->refcount++;
+	
+	return m;
+}
+
+void vine_mount_delete( struct vine_mount *m )
+{
+	if(!m) return;
+	vine_file_delete(m->file);
```

Comment:
Deleting a file when the count of mounts go away may lead to weird segfault bugs. E.g., users for whatever reason may keep a list with the references to `vine_file`.  Those references will be free() under them when the mounts go away.

Reply:
I think the refcounting works within the universe of file/task create/delete/connect. If the user does this:

```
file = vine_file_local(...)
task = vine_task_create(...)
vine_task_add_input(task, file,...)
```

Then at this point the file will have **two** references: one from the original file create, and as second one from vine_task_add_input.  So, then they can delete the file and task in any order, and will only go away on the last reference:

```
vine_task_delete(task);
vine_file_delete(file);
```

Likewise, if you clone a task, then you get a third reference to the underlying files.

Now, you are correct that it will break if the user keeps around multiple pointers to the same file.  To fix that, we need to move the file management into the manager object itself, and refer to them with unique identifiers.  We aren't there yet, but that's where I want to go...